### PR TITLE
Data reader

### PR DESCRIPTION
**Pull request by jsaarimaa**
Added functions to read data from the data section of a px file stream.
Added a new application for running performance benchmarks.

### CODE REVIEW

File: Px.Utils.TestingApp/InteractiveFlow.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+﻿using Px.Utils.TestingApp;
+using Px.Utils.TestingApp.Commands;
+
+namespace PxUtils.TestingApp
+{
+    /// <summary>
+    /// Runs the interactive flow of the application.
+    /// Propmpts the user for a commands and parameters and then runs the commands.
```
Comment (sakari-malkki) [inline Px.Utils.TestingApp/InteractiveFlow.cs:8]: Typo: "Propmpts"

**Review by sakari-malkki — DISMISSED** [review]
Apart from one summary typo couldn't find else.
There's some Sonar complaints about complex functions, but most of those take place in hot loops so fixing them might not be wise performance wise. The last one, related to the benchmark tool is 16, so changing that just to get it to <= 15 doesn't seem worthwhile either.

File: Px.Utils.TestingApp/InteractiveFlow.cs (hunk 1 above)
Comment (jsaarimaa) [reply Px.Utils.TestingApp/InteractiveFlow.cs]: Fixed

File: Px.Utils/PxFile/Data/DataValueParsers.cs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        /// <param name="buffer">The characters to parse.</param>
+        /// <param name="len">The number of characters to parse.</param>
+        /// <returns>A <see cref="DoubleDataValue"/> instance representing the parsed value.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static DoubleDataValue FastParseDoubleDataValueDangerous(char[] buffer, int len)
+        {
+            if (buffer[0] == '"')
+            {
+                if (buffer[1] == '-') return new DoubleDataValue(0, DataValueType.Nill);
+                return new DoubleDataValue(0, (DataValueType)(len - 2));
```
Comment (reenari) [inline Px.Utils/PxFile/Data/DataValueParsers.cs:24]: I'm not fan of comments but this needs a comment, both here and in DataValueType

File: Px.Utils/PxFile/Data/StreamUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            byte[] buffer = new byte[bufferSize];
+
+            long read;
+            int keywordIndex = 0;
+            bool searchMode = true;
+
+            do
+            {
+                read = stream.Read(buffer, 0, bufferSize);
+                long indexInBuffer = FindInBuffer(read, buffer, keywordBytes, entrySeparator, ref searchMode, ref keywordIndex);
```
Comment (reenari) [inline Px.Utils/PxFile/Data/StreamUtilities.cs:33]: What if the keyword lands between two buffer reads ?

File: Px.Utils.UnitTests/PxFileTests/DataTests/StreamUtilitiesTests.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+            // Assert
+            Assert.AreEqual(-1, position);
+        }
+
+        [TestMethod]
+        public void FindKeywordTest_DataKeywordInTheMiddleOfStream_ReturnsIndex()
+        {
+            // Arrange
+            byte[] data = Encoding.UTF8.GetBytes("DATA=123");
```
Comment (reenari) [inline Px.Utils.UnitTests/PxFileTests/DataTests/StreamUtilitiesTests.cs:47]: DATA keyword is not in the middle of the stream in this test case

File: Px.Utils/PxFile/Data/StreamUtilities.cs (hunk 3 above)
Comment (jsaarimaa) [reply Px.Utils/PxFile/Data/StreamUtilities.cs]: Refactored the code so this is no longer an issue. Added a unit test to confirm that this is no longer an issue.

File: Px.Utils/PxFile/Data/DataValueParsers.cs (hunk 2 above)
Comment (jsaarimaa) [reply Px.Utils/PxFile/Data/DataValueParsers.cs:24]: Added comments.

File: Px.Utils.UnitTests/PxFileTests/DataTests/StreamUtilitiesTests.cs (hunk 4 above)
Comment (jsaarimaa) [reply Px.Utils.UnitTests/PxFileTests/DataTests/StreamUtilitiesTests.cs]: Fixed the test to match the description.

**Review by lehtu — APPROVED** [review]

**Review by lehtu — APPROVED** [review]

**Pull request merged**